File: src/main/kotlin/com/huawei/cangjie1/psi/CjParameterList.java
package com.huawei.cangjie1.psi;

import com.huawei.cangjie1.lexer.CjTokens;
import com.huawei.cangjie1.psi.stubs.CangJiePlaceHolderStub;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import com.huawei.cangjie1.psi.psiUtil.CjPsiUtilKt;
import java.util.List;
import com.huawei.cangjie1.psi.stubs.elements.CjStubElementTypes;

public class CjParameterList extends CjElementImplStub<CangJiePlaceHolderStub<CjParameterList>> {
    public CjParameterList(@NotNull ASTNode node) {
        super(node);
    }

    public CjParameterList(@NotNull CangJiePlaceHolderStub<CjParameterList> stub) {
        super(stub, CjStubElementTypes.VALUE_PARAMETER_LIST);
    }


    @Override
    public String toString() {
        return getNode().getElementType().toString();
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, D data) {
        return visitor.visitParameterList(this, data);
    }

    @Override
    public PsiElement getParent() {
        CangJiePlaceHolderStub<CjParameterList> stub = getStub();
        return stub != null ? stub.getParentStub().getPsi() : super.getParent();
    }

    @NotNull
    public List<CjParameter> getParameters() {
        return getStubOrPsiChildrenAsList(CjStubElementTypes.VALUE_PARAMETER);
    }

    @NotNull
    public CjParameter addParameter(@NotNull CjParameter parameter) {
        return EditCommaSeparatedListHelper.INSTANCE.addItem(this, getParameters(), parameter);
    }

    @NotNull
    public CjParameter addParameterBefore(@NotNull CjParameter parameter, @Nullable CjParameter anchor) {
        return EditCommaSeparatedListHelper.INSTANCE.addItemBefore(this, getParameters(), parameter, anchor);
    }

    @NotNull
    public CjParameter addParameterAfter(@NotNull CjParameter parameter, @Nullable CjParameter anchor) {
        return EditCommaSeparatedListHelper.INSTANCE.addItemAfter(this, getParameters(), parameter, anchor);
    }

    public void removeParameter(@NotNull CjParameter parameter) {
        EditCommaSeparatedListHelper.INSTANCE.removeItem(parameter);
    }

    public void removeParameter(int index) {
        removeParameter(getParameters().get(index));
    }

    public CjDeclarationWithBody getOwnerFunction() {
        PsiElement parent = getParentByStub();
        if (!(parent instanceof CjDeclarationWithBody)) return null;
        return (CjDeclarationWithBody) parent;
    }

    @Nullable
    public PsiElement getRightParenthesis() {
        return findChildByType(CjTokens.RPAR);
    }

    @Nullable
    public PsiElement getLeftParenthesis() {
        return findChildByType(CjTokens.LPAR);
    }

    @Nullable
    public PsiElement getFirstComma() {
        return findChildByType(CjTokens.COMMA);
    }

    @Nullable
    public PsiElement getTrailingComma() {
        PsiElement parentElement = getParent();
//        if (parentElement instanceof CjFunctionLiteral || parentElement instanceof CjPropertyAccessor) {
//            return CjPsiUtilKt.getTrailingCommaByElementsList(this);
//        } else {
            return CjPsiUtilKt.getTrailingCommaByClosingElement(getRightParenthesis());
//        }
    }
}

File: psi/src/main/kotlin/org/cangnova/cangjie/psi/CjParameterList.kt
/*
 * Copyright 2026 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */

package org.cangnova.cangjie.psi

import org.cangnova.cangjie.lexer.CjTokens
import org.cangnova.cangjie.psi.EditCommaSeparatedListHelper.addItem
import org.cangnova.cangjie.psi.EditCommaSeparatedListHelper.addItemAfter
import org.cangnova.cangjie.psi.EditCommaSeparatedListHelper.addItemBefore
import org.cangnova.cangjie.psi.EditCommaSeparatedListHelper.removeItem
import org.cangnova.cangjie.psi.psiUtil.getTrailingCommaByClosingElement
import org.cangnova.cangjie.psi.stubs.CangJiePlaceHolderStub
import org.cangnova.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement

class CjParameterList : CjElementImplStub<CangJiePlaceHolderStub<CjParameterList>> {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJiePlaceHolderStub<CjParameterList>) : super(stub, CjStubElementTypes.VALUE_PARAMETER_LIST)

    override fun toString(): String {
        return node.elementType.toString()
    }

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D): R? {
        return visitor.visitParameterList(this, data)
    }

    override fun getParent(): PsiElement? {
        val stub: CangJiePlaceHolderStub<CjParameterList>? = stub
        return if (stub != null) stub.parentStub.psi else super.getParent()
    }

    val parameters: List<CjParameter>
        get() = getStubOrPsiChildrenAsList(CjStubElementTypes.VALUE_PARAMETER)

    fun addParameter(parameter: CjParameter): CjParameter {
        return addItem(
            this,
            parameters,
            parameter,
        )
    }

    fun addParameterBefore(parameter: CjParameter, anchor: CjParameter?): CjParameter {
        return addItemBefore(
            this,
            parameters,
            parameter,
            anchor,
        )
    }

    fun addParameterAfter(parameter: CjParameter, anchor: CjParameter?): CjParameter {
        return addItemAfter(
            this,
            parameters,
            parameter,
            anchor,
        )
    }

    fun removeParameter(parameter: CjParameter) {
        removeItem(parameter)
    }

    fun removeParameter(index: Int) {
        removeParameter(parameters[index])
    }

    val ownerFunction: CjDeclarationWithBody?
        get() {
            val parent = parentByStub as? CjDeclarationWithBody ?: return null
            return parent
        }

    val rightParenthesis: PsiElement?
        get() = findChildByType(CjTokens.RPAR)

    val leftParenthesis: PsiElement?
        get() = findChildByType(CjTokens.LPAR)

    val firstComma: PsiElement?
        get() = findChildByType(CjTokens.COMMA)

    val trailingComma: PsiElement?
        get() {
            val parentElement = parent
            //        if (parentElement instanceof CjFunctionLiteral || parentElement instanceof CjPropertyAccessor) {
//            return CjPsiUtilKt.getTrailingCommaByElementsList(this);
//        } else {
            return getTrailingCommaByClosingElement(rightParenthesis)
            //        }
        }
}
